types/react-query/index.d.ts

```diff
@@ -1,142 +1,370 @@
+}
+
+export interface InfiniteQueryResult<TResult, TMoreVariable> extends QueryResultBase<TResult> {
+    data: TResult[];
     isFetchingMore: boolean;
-    canFetchMore: boolean;
-    fetchMore: (variables?: TVariables) => Promise<TResult>;
+    fetchMore: (moreVariable?: TMoreVariable | false) => Promise<TResult[]> | undefined;
 }
 
-export function prefetchQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: PrefetchQueryOptions<TResult>
-): Promise<TResult>;
+// export function prefetchQuery<TResult, TVariables extends object>(
+//     queryKey: QueryKey<TVariables>,
+//     queryFn: QueryFunction<TResult, TVariables>,
+//     options?: PrefetchQueryOptions<TResult>,
+// ): Promise<TResult>;
 
 export interface PrefetchQueryOptions<TResult> extends QueryOptions<TResult> {
     force?: boolean;
 }
 
-export function useMutation<TResults, TVariables extends object>(
+export function useMutation<TResults, TVariables = undefined>(
     mutationFn: MutationFunction<TResults, TVariables>,
-    mutationOptions?: MutationOptions
+    mutationOptions?: MutationOptions<TResults, TVariables>,
 ): [MutateFunction<TResults, TVariables>, MutationResult<TResults>];
 
-export type MutationFunction<TResults, TVariables extends object> = (
-    variables: TVariables,
-) => Promise<TResults>;
+export type MutationFunction<TResults, TVariables> = (variables: TVariables) => Promise<TResults>;
+
+export interface MutateOptions<TResult, TVariables> {
+    onSuccess?: (data: TResult, variables: TVariables) => Promise<void> | void;
+    onError?: (error: any, variables: TVariables, snapshotValue: unknown) => Promise<void> | void;
+    onSettled?: (
+        data: undefined | TResult,
+        error: unknown | null,
+        variables: TVariables,
+        snapshotValue?: unknown,
+    ) => Promise<void> | void;
+    throwOnError?: boolean;
+}
 
-export interface MutationOptions {
-    refetchQueries?: Array<string | [string, object]>;
-    refetchQueriesOnFailure?: boolean;
+export interface MutationOptions<TResult, TVariables> extends MutateOptions<TResult, TVariables> {
+    onMutate?: (variables: TVariables) => Promise<unknown> | unknown;
+    useErrorBoundary?: boolean;
 }
 
-export type MutateFunction<TResults, TVariables extends object> = (
-    variables?: TVariables,
-    options?: {
-        updateQuery?: string | [string, object],
-        waitForRefetchQueries?: boolean;
-    }
-) => Promise<TResults>;
+export type MutateFunction<TResult, TVariables> = undefined extends TVariables
+    ? (variables?: TVariables, options?: MutateOptions<TResult, TVariables>) => Promise<TResult>
+    : (variables: TVariables, options?: MutateOptions<TResult, TVariables>) => Promise<TResult>;
 
 export interface MutationResult<TResults> {
-    data: TResults | null;
-    isLoading: boolean;
-    error: null | Error;
+    status: 'idle' | 'loading' | 'error' | 'success';
```

Comment:
Can you add discrimanated untion here?

Reply:
Added. [Source code](https://github.com/tannerlinsley/react-query/blob/v1.1.4/src/useMutation.js#L29-L49) is even weirder there, it just not have `error`, `data` in some states. The definition now follows actual behaviour.